src/aviary/functional.py

```diff
@@ -0,0 +1,175 @@
+from collections.abc import Callable
+from functools import update_wrapper
+from typing import Any
+
+from pydantic import BaseModel, Field
+
+from aviary.env import Environment, Frame
+from aviary.message import Message
+from aviary.tools import Messages, Tool, ToolRequestMessage
+from aviary.utils import is_coroutine_callable
+
+
+class DynamicState(BaseModel):
+    """Dynamic env state model that adapts to provided extras."""
+
+    reward: float = 0
+    done: bool = False
+    extras: dict[str, Any] = Field(default_factory=dict)
```

Comment:
Can you add a `description` for `extras`

Mainly I feel our `extras` here semantically overlaps with Pydantic `ConfigDict(extra="allow")`, and it would be nice to mention it's the same meaning, just we made it explicitly an attribute

---

An aside is perhaps we should use `model_config = ConfigDict(extra="allow")`, and then we can remove `__getattr__`

Reply:
Added a comment - you cannot use `__getattr__` with `extra` because you get an infinite recursion (since it overrides `getattr` already), so I just added a new field called extras.